x-pack/plugins/enterprise_search/public/applications/app_search/components/meta_engine_creation/meta_engine_creation_logic.ts

```diff
@@ -0,0 +1,131 @@
+ * 2.0.
+ */
+
+import { generatePath } from 'react-router-dom';
+
+import { kea, MakeLogicType } from 'kea';
+
+import { Meta } from '../../../../../common/types';
+import { DEFAULT_META } from '../../../shared/constants';
+import { flashAPIErrors, setQueuedSuccessMessage } from '../../../shared/flash_messages';
+import { HttpLogic } from '../../../shared/http';
+import { KibanaLogic } from '../../../shared/kibana';
+import { ENGINE_PATH } from '../../routes';
+import { formatApiName } from '../../utils/format_api_name';
+import { EngineDetails } from '../engine/types';
+
+import { META_ENGINE_CREATION_SUCCESS_MESSAGE } from './constants';
+
+interface MetaEngineCreationValues {
+  indexedEngineNames: string[];
+  name: string;
+  rawName: string;
+  selectedIndexedEngineNames: string[];
+}
+
+interface MetaEngineCreationActions {
+  fetchIndexedEngineNames(page?: number): { page: number };
+  onEngineCreationSuccess(): void;
+  setIndexedEngineNames(
+    indexedEngineNames: MetaEngineCreationValues['indexedEngineNames']
+  ): { indexedEngineNames: MetaEngineCreationValues['indexedEngineNames'] };
+  setRawName(rawName: string): { rawName: string };
+  setSelectedIndexedEngineNames(
+    selectedIndexedEngineNames: MetaEngineCreationValues['selectedIndexedEngineNames']
+  ): { selectedIndexedEngineNames: MetaEngineCreationValues['selectedIndexedEngineNames'] };
+  submitEngine(): void;
+}
+
+export const MetaEngineCreationLogic = kea<
+  MakeLogicType<MetaEngineCreationValues, MetaEngineCreationActions>
+>({
+  path: ['enterprise_search', 'app_search', 'meta_engine_creation_logic'],
+  actions: {
+    fetchIndexedEngineNames: (page = DEFAULT_META.page.current) => ({ page }),
+    onEngineCreationSuccess: true,
+    setIndexedEngineNames: (indexedEngineNames) => ({ indexedEngineNames }),
+    setRawName: (rawName) => ({ rawName }),
+    setSelectedIndexedEngineNames: (selectedIndexedEngineNames) => ({ selectedIndexedEngineNames }),
+    submitEngine: () => null,
+  },
+  reducers: {
+    indexedEngineNames: [
+      [],
+      {
+        setIndexedEngineNames: (_, { indexedEngineNames }) => indexedEngineNames,
+      },
+    ],
+    rawName: [
+      '',
+      {
+        setRawName: (_, { rawName }) => rawName,
+      },
+    ],
+    selectedIndexedEngineNames: [
+      [],
+      {
+        setSelectedIndexedEngineNames: (_, { selectedIndexedEngineNames }) =>
+          selectedIndexedEngineNames,
+      },
+    ],
+  },
+  selectors: ({ selectors }) => ({
+    name: [() => [selectors.rawName], (rawName: string) => formatApiName(rawName)],
+  }),
+  listeners: ({ values, actions }) => ({
+    fetchIndexedEngineNames: async ({ page }) => {
+      const { http } = HttpLogic.values;
+      let response;
+
+      try {
+        response = (await http.get('/api/app_search/engines', {
+          query: { type: 'indexed', 'page[current]': page, 'page[size]': DEFAULT_META.page.size },
+        })) as { results: EngineDetails[]; meta: Meta };
```

Comment:
I think you can do this instead typing wise if you don't want to cast `as`

```suggestion
      let response: { results: EngineDetails[]; meta: Meta } | undefined;

      try {
        response = await http.get('/api/app_search/engines', {
          query: { type: 'indexed', 'page[current]': page, 'page[size]': DEFAULT_META.page.size },
        });
```